### PR TITLE
Fixing timezone option of formatDateTime

### LINKED ISSUE #72877: formatDateTime %z bug: extra random crufty characters instead of +
**Company or project name**

Cribl

**Describe what's wrong**

When using `formatDateTime()` with a format that contains `%z` mixed with other placeholders, you get extra random crufty characters instead of `+`.

Easily reproduced:
https://fiddle.clickhouse.com/4e868ff7-89fd-4bbd-891a-b2b0af9e725f

**Does it reproduce on the most recent release?**

Yes

**How to reproduce**

See Fiddle link above.  Use `%z` in comination with pretty much any other format placeholder such as `%W`, i.e. 

```formatDateTime(ts, '%W%z')```

Doesn't matter which version/interface/settings.

**Expected behavior**

`%z` should yield `+0000` consistently with no random/crufty characters.

### PR DESCRIPTION
**Pull request by ollidraese**
The mysqlTimezoneOffset function, called for the formatDateTime(ts, '%W%z') scalar function only stores the '-' for negative timezone offsets but no '+' for positive offsets. The memory that is passed into this function is not necessarily initialized, so the function skips the character for the +/- and only writes the offset value. This leads to the corrupted output as described in the following issue:

Closes https://github.com/ClickHouse/ClickHouse/issues/72877

### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in an official stable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix positive timezone formatting of formatDateTime scalar function.

### Documentation entry for user-facing changes

n/a

#### CI Settings (Only check the boxes if you know what you are doing):
- [ ]  Allow: All Required Checks
- [ ]  Allow: Stateless tests
- [ ]  Allow: Stateful tests
- [ ]  Allow: Integration Tests
- [ ]  Allow: Performance tests
- [ ]  Allow: All Builds
- [ ]  Allow: batch 1, 2 for multi-batch jobs
- [ ]  Allow: batch 3, 4, 5, 6 for multi-batch jobs
---
- [ ]  Exclude: Style check
- [ ]  Exclude: Fast test
- [ ]  Exclude: All with ASAN
- [ ]  Exclude: All with TSAN, MSAN, UBSAN, Coverage
- [ ]  Exclude: All with aarch64, release, debug
---
- [ ]  Run only fuzzers related jobs (libFuzzer fuzzers, AST fuzzers, etc.)
- [ ]  Exclude: AST fuzzers
---
- [ ]  Do not test
- [ ]  Woolen Wolfdog
- [ ]  Upload binaries for special builds
- [ ]  Disable merge-commit
- [ ]  Disable CI cache

### CODE REVIEW

**Comment by robot-ch-test-poll4** [comment]
*This is an automated comment for commit 07c8c772e7f4c499f2606f459660dba1843ab59c with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/73091/07c8c772e7f4c499f2606f459660dba1843ab59c/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Docs check</td><td>Builds and tests the documentation</td><td>✅ success</td></tr>
<tbody>
</table>
</details>

**Comment by alexkats** [comment]
Thanks for the change, please add a test for it

**Review by alexkats — CHANGES_REQUESTED** [review]
Thanks for the change, please add a test for it. You can use something similar to reproduction in the issue and put you test in `tests/queries/0_stateless`

**Comment by alexkats** [comment]
Also style check has failed because of trailing spaces

File: tests/queries/0_stateless/00718_format_datetime_1.sql [hunk 1]
```diff
@@ -3,3 +3,4 @@ select formatDateTime(toDateTime64('1962-12-08 18:11:29.123', 3, 'UTC'), '%F %T.
 select formatDateTime(toDateTime64('1969-12-31 23:59:59.999', 3, 'UTC'), '%F %T.%f');
 select formatDateTime(toDateTime64('1970-01-01 00:00:00.000', 3, 'UTC'), '%F %T.%f');
 select formatDateTime(toDateTime64('1970-01-01 00:00:00.001', 3, 'UTC'), '%F %T.%f');
+select formatDateTime(toDateTime64('2024-12-11 12:34:56.000', 3, 'UTC'), '%W%z');
```
Comment (alexkats) [inline tests/queries/0_stateless/00718_format_datetime_1.sql:6]: In general new test cases should be created in a new file. Let's do it and add a couple more tests with different time zones (both negative and positive), so we'll have this check for all the cases in the future

**Comment by ollidraese** [comment]
@alexkats I moved the tests (now multiple timezones) into a new file. The `clickhouse-test` script is failing for me (across different macos machines), so, this might be a script regression. I opened https://github.com/ClickHouse/ClickHouse/issues/73310 for this. Thx.

**Review by alexkats — APPROVED** [review]
LGTM, thanks again for the changes. Waiting for CI to finish.

**Comment by alexkats** [comment]
> @alexkats I moved the tests (now multiple timezones) into a new file. The `clickhouse-test` script is failing for me (across different macos machines), so, this might be a script regression. I opened https://github.com/ClickHouse/ClickHouse/issues/73310 for this. Thx.

I'll try to take a look at the script later on.

**Comment by alexkats** [comment]
Taking a brief look, all CI failures seem unrelated to the changes. Will take a closer look later

**Pull request merged**